depends: Use Qt top-level build facilities

hebasto opened the pull request:
This PR:
- uses Qt top-level build facilities without the need to download all-in-one archive
- is based on **BlockMechanic**'s [idea](https://github.com/bitcoin/bitcoin/pull/20600), and is an alternative to #20600
- makes it easy to integrate [new modules](https://github.com/bitcoin/bitcoin/pull/16883) into static builds
- has the minimal diff
- makes the qt package build process streamlined by dropping some patches and hacks (an alternative to  #21420 and #20642)

Fixes #18536 (a non-intrusive alternative to #21589 and #19785).

Fixes #14648.

Fixes #21588 (a non-intrusive alternative to #21591).

Required for adding [Wayland support](https://github.com/bitcoin/bitcoin/issues/19950) on Linux.

---

**Note for reviewers**: With 9046de8a4cbc3899fed9eae084115f423e7ac5bd from #21995 it is easy to verify that there are no changes in the resulted `qt` package archive on the per commit basis. For example, for `HOST=i686-pc-linux-gnu` no commit in this PR introduces any changes.

BlockMechanic commented:
This seems pretty clean, testing. Had you tested this with additional modules ? (that was my reasoning for changing the way QT builds)

hebasto commented:
> Had you tested this with additional modules ? (that was my reasoning for changing the way QT builds)

I've started with the following `qt.pro`
```
# Create the super cache so modules will add themselves to it.
cache(, super)

CONFIG += prepare_docs qt_docs_targets

TEMPLATE = subdirs
SUBDIRS = qtbase qtdeclarative qtgraphicaleffects qtquickcontrols2 qttools qttranslations

qtbase.target = module-qtbase

qtdeclarative.target = module-qtdeclarative
qtdeclarative.depends = qtbase

qtgraphicaleffects.target = module-qtgraphicaleffects
qtgraphicaleffects.depends = qtdeclarative

qtquickcontrols2.target = module-qtquickcontrols2
qtquickcontrols2.depends = qtgraphicaleffects

qttools.target = module-qttools
qttools.depends = qtbase qtdeclarative

qttranslations.target = module-qttranslations
qttranslations.depends = qttools

QT_SKIP_MODULES =

load(qt_configure)
```

that works, and reduced changes to this minimal diff.

BlockMechanic commented:
> > Had you tested this with additional modules ? (that was my reasoning for changing the way QT builds)
> 
> I've started with the following `qt.pro`
> 
> ```
> # Create the super cache so modules will add themselves to it.
> cache(, super)
> 
> CONFIG += prepare_docs qt_docs_targets
> 
> TEMPLATE = subdirs
> SUBDIRS = qtbase qtdeclarative qtgraphicaleffects qtquickcontrols2 qttools qttranslations
> 
> qtbase.target = module-qtbase
> 
> qtdeclarative.target = module-qtdeclarative
> qtdeclarative.depends = qtbase
> 
> qtgraphicaleffects.target = module-qtgraphicaleffects
> qtgraphicaleffects.depends = qtdeclarative
> 
> qtquickcontrols2.target = module-qtquickcontrols2
> qtquickcontrols2.depends = qtgraphicaleffects
> 
> qttools.target = module-qttools
> qttools.depends = qtbase qtdeclarative
> 
> qttranslations.target = module-qttranslations
> qttranslations.depends = qttools
> 
> QT_SKIP_MODULES =
> 
> load(qt_configure)
> ```
> 
> that works, and reduced changes to this minimal diff.

I will test this, great work !

promag commented:
Is there any advantage of this approach beside avoid downloading the full source archive?

hebasto commented:
> Is there any advantage of this approach beside avoid downloading the full source archive?

Not dealing with unneeded stuff, I guess.

BlockMechanic commented:
> Is there any advantage of this approach beside avoid downloading the full source archive?

Not that i can see. but it is much cleaner IMO.

icota commented on depends/patches/qt/qt.pro:
```diff
@@ -0,0 +1,17 @@
+# Create the super cache so modules will add themselves to it.
```
Should we move this file and `mac-qmake.conf` to `../configs/qt`?

hebasto replied:
Files in the `depends/patches/<package>/` directory are subjects of a [build hash calculation](https://github.com/bitcoin/bitcoin/blob/25bc840e038f5cc3a1d70d6fbaf1dc2eb2c6e460/depends/funcs.mk#L37-L40). So, I don't think that moving some of them to another directory is a good idea.

icota replied:
One could add `depends/configs` to the calculation. No big deal either way, perhaps I'm nitpicking

laanwj commented:
> Is there any advantage of this approach beside avoid downloading the full source archive?

The full Qt source archive is **huge**. I prefer being selective.

hebasto commented:
Rebased 770e40c4af3e77dcc9f0cc1fa4973d4f9a1baf5a -> b72ff7c3c30a41167f3a14f89ad18f6cf7124122 ([pr20641.01](https://github.com/hebasto/bitcoin/commits/pr20641.01) -> [pr20641.02](https://github.com/hebasto/bitcoin/commits/pr20641.02)) due to the conflict with #20673.

promag commented:
> > Is there any advantage of this approach beside avoid downloading the full source archive?
> 
> The full Qt source archive is **huge**. I prefer being selective.

@laanwj true, in Qt 5.15. But in Qt 6 the fat module (qtwebengine-everywhere-src-5.15.2.tar.xz with 267M) is gone, see
http://download.qt.io/official_releases/qt/5.15/5.15.2/single/ and http://download.qt.io/official_releases/qt/6.0/6.0.0/single/.

promag commented:
Concept ACK.

BlockMechanic commented:
@hebasto check out how I selected the required components using gitmodules, it effectively keeps compile time the same, while allowing easier addition/removal of modules.

hebasto commented:
> @hebasto check out how I selected the required components using gitmodules, it effectively keeps compile time the same, while allowing easier addition/removal of modules.

Which branch?

BlockMechanic commented:
> > @hebasto check out how I selected the required components using gitmodules, it effectively keeps compile time the same, while allowing easier addition/removal of modules.
> 
> Which branch?
see files gitmodules and fix_git_modules.patch here :- https://github.com/bitcoin/bitcoin/pull/20600/files

hebasto committed:
build: Use Qt top-level build facilities

hebasto commented:
Rebased, improved, OP updated,

hebasto commented:
#### Guix builds:
```
$ find guix-build-$(git rev-parse --short=12 HEAD)/output/ -type f -print0 | env LC_ALL=C sort -z | xargs -r0 sha256sum
1666b7e6a342b0089a900edb02721dcede74269eaa2976e5a01f443c0c0929fc  guix-build-5ca6e8f16bc8/output/aarch64-linux-gnu/bitcoin-5ca6e8f16bc8-aarch64-linux-gnu-debug.tar.gz
bcd89e2835e7b25b84c1e32f81e8424edd0be22301ae44b778648a3ad3e40891  guix-build-5ca6e8f16bc8/output/aarch64-linux-gnu/bitcoin-5ca6e8f16bc8-aarch64-linux-gnu.tar.gz
2186034213a86410ee03330ef50c131808bcbcb145643b39dcbd7d975534402a  guix-build-5ca6e8f16bc8/output/aarch64-linux-gnu/inputs.SHA256SUMS
4502fa86a010e2fe06dbe973240039dc392e5c862dce191f2e67544123b89861  guix-build-5ca6e8f16bc8/output/arm-linux-gnueabihf/bitcoin-5ca6e8f16bc8-arm-linux-gnueabihf-debug.tar.gz
5962d849b9b6370c5f2a0e1b77866e652ad9676fc8abcee614131e1e31b96e40  guix-build-5ca6e8f16bc8/output/arm-linux-gnueabihf/bitcoin-5ca6e8f16bc8-arm-linux-gnueabihf.tar.gz
2186034213a86410ee03330ef50c131808bcbcb145643b39dcbd7d975534402a  guix-build-5ca6e8f16bc8/output/arm-linux-gnueabihf/inputs.SHA256SUMS
e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855  guix-build-5ca6e8f16bc8/output/dist-archive/SKIPATTEST.TAG
32bb3c127d84add96715480970d97c0a1e76651a5ad4c6c245af2b479b913710  guix-build-5ca6e8f16bc8/output/dist-archive/bitcoin-5ca6e8f16bc8.tar.gz
691a14242a62adbffd4e6b8b805b0122826edcde46912180b11b6413f531408e  guix-build-5ca6e8f16bc8/output/powerpc64-linux-gnu/bitcoin-5ca6e8f16bc8-powerpc64-linux-gnu-debug.tar.gz
324747b82555543b1c2a263c8f9ac60a8e3d183ddba7e62d0ff49bae3be75706  guix-build-5ca6e8f16bc8/output/powerpc64-linux-gnu/bitcoin-5ca6e8f16bc8-powerpc64-linux-gnu.tar.gz
2186034213a86410ee03330ef50c131808bcbcb145643b39dcbd7d975534402a  guix-build-5ca6e8f16bc8/output/powerpc64-linux-gnu/inputs.SHA256SUMS
dd4939e4fd6a244100ed58a4d479e126f7829b25995a829fe032e4f9e3923d4c  guix-build-5ca6e8f16bc8/output/powerpc64le-linux-gnu/bitcoin-5ca6e8f16bc8-powerpc64le-linux-gnu-debug.tar.gz
611c8443ddc2be9da5301b787fd8597b397d1b5a9c17fb408337880315abcdef  guix-build-5ca6e8f16bc8/output/powerpc64le-linux-gnu/bitcoin-5ca6e8f16bc8-powerpc64le-linux-gnu.tar.gz
2186034213a86410ee03330ef50c131808bcbcb145643b39dcbd7d975534402a  guix-build-5ca6e8f16bc8/output/powerpc64le-linux-gnu/inputs.SHA256SUMS
50f06f12b4a5b34f13e27f364307ec9898014d13c62b49ac460ac54c4ba833e8  guix-build-5ca6e8f16bc8/output/riscv64-linux-gnu/bitcoin-5ca6e8f16bc8-riscv64-linux-gnu-debug.tar.gz
a4fdf95b94cbe6d88337997640e6dd2bbcfab0abce6d9c80006206832f9f65c3  guix-build-5ca6e8f16bc8/output/riscv64-linux-gnu/bitcoin-5ca6e8f16bc8-riscv64-linux-gnu.tar.gz
2186034213a86410ee03330ef50c131808bcbcb145643b39dcbd7d975534402a  guix-build-5ca6e8f16bc8/output/riscv64-linux-gnu/inputs.SHA256SUMS
2c0054cb1d0e4f740d0bc2660c463e909b3fd008c03b1b3602c0bb267402152c  guix-build-5ca6e8f16bc8/output/x86_64-apple-darwin18/bitcoin-5ca6e8f16bc8-osx-unsigned.dmg
0625e91eee410e2e0e75d3131dafaed2f50745dafd7f602a1b52c4620f8c5e7f  guix-build-5ca6e8f16bc8/output/x86_64-apple-darwin18/bitcoin-5ca6e8f16bc8-osx-unsigned.tar.gz
981835f2c06636539f2fff39f4baef4e198386d7489ac3775300d15e60da3d46  guix-build-5ca6e8f16bc8/output/x86_64-apple-darwin18/bitcoin-5ca6e8f16bc8-osx64.tar.gz
2186034213a86410ee03330ef50c131808bcbcb145643b39dcbd7d975534402a  guix-build-5ca6e8f16bc8/output/x86_64-apple-darwin18/inputs.SHA256SUMS
fe481c8197799ffdc2415f5f2ec567ffbf411541d0f690e38df529a3a85d829b  guix-build-5ca6e8f16bc8/output/x86_64-linux-gnu/bitcoin-5ca6e8f16bc8-x86_64-linux-gnu-debug.tar.gz
dea06c3d3338fc6ace7e0c8b8f1e4786e558031cf69171f50a760fd80dfbf18a  guix-build-5ca6e8f16bc8/output/x86_64-linux-gnu/bitcoin-5ca6e8f16bc8-x86_64-linux-gnu.tar.gz
2186034213a86410ee03330ef50c131808bcbcb145643b39dcbd7d975534402a  guix-build-5ca6e8f16bc8/output/x86_64-linux-gnu/inputs.SHA256SUMS
10890f11d650bce83dc0faed59177896f9ff09f5a4ff0a973061342825f25d40  guix-build-5ca6e8f16bc8/output/x86_64-w64-mingw32/bitcoin-5ca6e8f16bc8-win-unsigned.tar.gz
813a685afe39bb8a41885dbc05293aca34a78734e40f99df2a75d89f4c5eadfd  guix-build-5ca6e8f16bc8/output/x86_64-w64-mingw32/bitcoin-5ca6e8f16bc8-win64-debug.zip
5b938ebab7f9a1f46bcf32c6c759487e1495f78f55431bd89bd77a531fe3ae63  guix-build-5ca6e8f16bc8/output/x86_64-w64-mingw32/bitcoin-5ca6e8f16bc8-win64-setup-unsigned.exe
4270ea718b1cd9b095917b42cc47dcc97b436ae5fc2dbe6585cdd059d65a5bfb  guix-build-5ca6e8f16bc8/output/x86_64-w64-mingw32/bitcoin-5ca6e8f16bc8-win64.zip
2186034213a86410ee03330ef50c131808bcbcb145643b39dcbd7d975534402a  guix-build-5ca6e8f16bc8/output/x86_64-w64-mingw32/inputs.SHA256SUMS
```

hebasto commented:
#### Gitian builds:

- Linux
```
Generating report
36b88f354e1354df4d350a568a460f1068c68730eb249476b35df29ee1859d3f  bitcoin-5ca6e8f16bc8-aarch64-linux-gnu-debug.tar.gz
6becaea105d391be157e56b313501bd8fcbb7eb739b06fe44008577a2edf02ae  bitcoin-5ca6e8f16bc8-aarch64-linux-gnu.tar.gz
b26eccbb6536f64fa0d29005d41bb95bad4087aaf134543377d07eab14de26b2  bitcoin-5ca6e8f16bc8-arm-linux-gnueabihf-debug.tar.gz
6e3765b74d1763576129ca06121eaeef00f96a1ca25fdca6b662a76b69c6e7ff  bitcoin-5ca6e8f16bc8-arm-linux-gnueabihf.tar.gz
731e135a3590afade51ba37bc0783f36773da640a296fee290f99b04215c8b78  bitcoin-5ca6e8f16bc8-powerpc64-linux-gnu-debug.tar.gz
4baccc8e005ee0a8bfa2086cea88deb049bb8949305e2e89a1a390f602f28365  bitcoin-5ca6e8f16bc8-powerpc64-linux-gnu.tar.gz
ba927bfe17f3687544b57e0f7139c8ce1c35a48db902e2dbf701bd524ba32950  bitcoin-5ca6e8f16bc8-powerpc64le-linux-gnu-debug.tar.gz
f80a2ce394f3e8cabc80611dd20a88069b816816233c2ee7cf8851e8bed396b6  bitcoin-5ca6e8f16bc8-powerpc64le-linux-gnu.tar.gz
220e2b5485ae2717436b00064f429ed46b08bf79a8e6aaabfb7a835dae7d9a06  bitcoin-5ca6e8f16bc8-riscv64-linux-gnu-debug.tar.gz
ca8966a635342107318cf894da603eaf4e690b295641af964b1a08c6f9fbea95  bitcoin-5ca6e8f16bc8-riscv64-linux-gnu.tar.gz
b4dc964183bc4d52bae9faad0f6b3ea8bf24f62d5c560dfa01bf95c8294eb03b  bitcoin-5ca6e8f16bc8-x86_64-linux-gnu-debug.tar.gz
5a4ed2710769e89b87518183275f95cea70091aca00f61a86296c12d7ba4f0c0  bitcoin-5ca6e8f16bc8-x86_64-linux-gnu.tar.gz
32bb3c127d84add96715480970d97c0a1e76651a5ad4c6c245af2b479b913710  src/bitcoin-5ca6e8f16bc8.tar.gz
f4a9492ca82680b9ce173b827657d37a9f25fa71006bb68a12667920052f43be  bitcoin-core-linux-22-res.yml
Done.
```

- Windows:
```
Generating report
cc7d18add2e21bb0ef71a15ea98dd7d1918cffe6808a3ab6dba85a887a2ccbfb  bitcoin-5ca6e8f16bc8-win-unsigned.tar.gz
e9c272519bdff9814ca26620dbeecae8bd9bc5e24b127cae85e92ae5374957f4  bitcoin-5ca6e8f16bc8-win64-debug.zip
d9062d2f6569430f641c3eefff4d3dd0ef29c81a92f2d424eb516a415367758b  bitcoin-5ca6e8f16bc8-win64-setup-unsigned.exe
f279c79ce6644a63d9c0be1e5037166eee002bda8844be47ae11a395b8c2ad36  bitcoin-5ca6e8f16bc8-win64.zip
32bb3c127d84add96715480970d97c0a1e76651a5ad4c6c245af2b479b913710  src/bitcoin-5ca6e8f16bc8.tar.gz
557f7613cc28706be546347cb2b49c5a1f3625864726be31df9db14d06f9b9ac  bitcoin-core-win-22-res.yml
Done.
```

- macOS:
```
Generating report
499c91afe1e932985ac4a1e2fd4ba43d25351c1a22f6de3ffd96be8f143ffc36  bitcoin-5ca6e8f16bc8-osx-unsigned.dmg
5bc6984dd7c4ab28b05f38021b841fdf46b8b68f0cefee7f54ac2c84dd86cd31  bitcoin-5ca6e8f16bc8-osx-unsigned.tar.gz
ff5ce0814c051dfb42ae6036a98087180c4423dc069168e819d03f1033c1a3a6  bitcoin-5ca6e8f16bc8-osx64.tar.gz
32bb3c127d84add96715480970d97c0a1e76651a5ad4c6c245af2b479b913710  src/bitcoin-5ca6e8f16bc8.tar.gz
b0c798594e9b6d29a9b91d3325157612c1f10ba3450d64ea5846bc5da5cb79a7  bitcoin-core-osx-22-res.yml
Done.
```

fanquake commented:
Guix builds:
```bash
bash-5.1# find guix-build-$(git rev-parse --short=12 HEAD)/output/ -type f -print0 | env LC_ALL=C sort -z | xargs -r0 sha256sum
1666b7e6a342b0089a900edb02721dcede74269eaa2976e5a01f443c0c0929fc  guix-build-5ca6e8f16bc8/output/aarch64-linux-gnu/bitcoin-5ca6e8f16bc8-aarch64-linux-gnu-debug.tar.gz
bcd89e2835e7b25b84c1e32f81e8424edd0be22301ae44b778648a3ad3e40891  guix-build-5ca6e8f16bc8/output/aarch64-linux-gnu/bitcoin-5ca6e8f16bc8-aarch64-linux-gnu.tar.gz
2186034213a86410ee03330ef50c131808bcbcb145643b39dcbd7d975534402a  guix-build-5ca6e8f16bc8/output/aarch64-linux-gnu/inputs.SHA256SUMS
4502fa86a010e2fe06dbe973240039dc392e5c862dce191f2e67544123b89861  guix-build-5ca6e8f16bc8/output/arm-linux-gnueabihf/bitcoin-5ca6e8f16bc8-arm-linux-gnueabihf-debug.tar.gz
5962d849b9b6370c5f2a0e1b77866e652ad9676fc8abcee614131e1e31b96e40  guix-build-5ca6e8f16bc8/output/arm-linux-gnueabihf/bitcoin-5ca6e8f16bc8-arm-linux-gnueabihf.tar.gz
2186034213a86410ee03330ef50c131808bcbcb145643b39dcbd7d975534402a  guix-build-5ca6e8f16bc8/output/arm-linux-gnueabihf/inputs.SHA256SUMS
e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855  guix-build-5ca6e8f16bc8/output/dist-archive/SKIPATTEST.TAG
32bb3c127d84add96715480970d97c0a1e76651a5ad4c6c245af2b479b913710  guix-build-5ca6e8f16bc8/output/dist-archive/bitcoin-5ca6e8f16bc8.tar.gz
691a14242a62adbffd4e6b8b805b0122826edcde46912180b11b6413f531408e  guix-build-5ca6e8f16bc8/output/powerpc64-linux-gnu/bitcoin-5ca6e8f16bc8-powerpc64-linux-gnu-debug.tar.gz
324747b82555543b1c2a263c8f9ac60a8e3d183ddba7e62d0ff49bae3be75706  guix-build-5ca6e8f16bc8/output/powerpc64-linux-gnu/bitcoin-5ca6e8f16bc8-powerpc64-linux-gnu.tar.gz
2186034213a86410ee03330ef50c131808bcbcb145643b39dcbd7d975534402a  guix-build-5ca6e8f16bc8/output/powerpc64-linux-gnu/inputs.SHA256SUMS
dd4939e4fd6a244100ed58a4d479e126f7829b25995a829fe032e4f9e3923d4c  guix-build-5ca6e8f16bc8/output/powerpc64le-linux-gnu/bitcoin-5ca6e8f16bc8-powerpc64le-linux-gnu-debug.tar.gz
611c8443ddc2be9da5301b787fd8597b397d1b5a9c17fb408337880315abcdef  guix-build-5ca6e8f16bc8/output/powerpc64le-linux-gnu/bitcoin-5ca6e8f16bc8-powerpc64le-linux-gnu.tar.gz
2186034213a86410ee03330ef50c131808bcbcb145643b39dcbd7d975534402a  guix-build-5ca6e8f16bc8/output/powerpc64le-linux-gnu/inputs.SHA256SUMS
50f06f12b4a5b34f13e27f364307ec9898014d13c62b49ac460ac54c4ba833e8  guix-build-5ca6e8f16bc8/output/riscv64-linux-gnu/bitcoin-5ca6e8f16bc8-riscv64-linux-gnu-debug.tar.gz
a4fdf95b94cbe6d88337997640e6dd2bbcfab0abce6d9c80006206832f9f65c3  guix-build-5ca6e8f16bc8/output/riscv64-linux-gnu/bitcoin-5ca6e8f16bc8-riscv64-linux-gnu.tar.gz
2186034213a86410ee03330ef50c131808bcbcb145643b39dcbd7d975534402a  guix-build-5ca6e8f16bc8/output/riscv64-linux-gnu/inputs.SHA256SUMS
2c0054cb1d0e4f740d0bc2660c463e909b3fd008c03b1b3602c0bb267402152c  guix-build-5ca6e8f16bc8/output/x86_64-apple-darwin18/bitcoin-5ca6e8f16bc8-osx-unsigned.dmg
0625e91eee410e2e0e75d3131dafaed2f50745dafd7f602a1b52c4620f8c5e7f  guix-build-5ca6e8f16bc8/output/x86_64-apple-darwin18/bitcoin-5ca6e8f16bc8-osx-unsigned.tar.gz
981835f2c06636539f2fff39f4baef4e198386d7489ac3775300d15e60da3d46  guix-build-5ca6e8f16bc8/output/x86_64-apple-darwin18/bitcoin-5ca6e8f16bc8-osx64.tar.gz
2186034213a86410ee03330ef50c131808bcbcb145643b39dcbd7d975534402a  guix-build-5ca6e8f16bc8/output/x86_64-apple-darwin18/inputs.SHA256SUMS
fe481c8197799ffdc2415f5f2ec567ffbf411541d0f690e38df529a3a85d829b  guix-build-5ca6e8f16bc8/output/x86_64-linux-gnu/bitcoin-5ca6e8f16bc8-x86_64-linux-gnu-debug.tar.gz
dea06c3d3338fc6ace7e0c8b8f1e4786e558031cf69171f50a760fd80dfbf18a  guix-build-5ca6e8f16bc8/output/x86_64-linux-gnu/bitcoin-5ca6e8f16bc8-x86_64-linux-gnu.tar.gz
2186034213a86410ee03330ef50c131808bcbcb145643b39dcbd7d975534402a  guix-build-5ca6e8f16bc8/output/x86_64-linux-gnu/inputs.SHA256SUMS
10890f11d650bce83dc0faed59177896f9ff09f5a4ff0a973061342825f25d40  guix-build-5ca6e8f16bc8/output/x86_64-w64-mingw32/bitcoin-5ca6e8f16bc8-win-unsigned.tar.gz
813a685afe39bb8a41885dbc05293aca34a78734e40f99df2a75d89f4c5eadfd  guix-build-5ca6e8f16bc8/output/x86_64-w64-mingw32/bitcoin-5ca6e8f16bc8-win64-debug.zip
5b938ebab7f9a1f46bcf32c6c759487e1495f78f55431bd89bd77a531fe3ae63  guix-build-5ca6e8f16bc8/output/x86_64-w64-mingw32/bitcoin-5ca6e8f16bc8-win64-setup-unsigned.exe
4270ea718b1cd9b095917b42cc47dcc97b436ae5fc2dbe6585cdd059d65a5bfb  guix-build-5ca6e8f16bc8/output/x86_64-w64-mingw32/bitcoin-5ca6e8f16bc8-win64.zip
2186034213a86410ee03330ef50c131808bcbcb145643b39dcbd7d975534402a  guix-build-5ca6e8f16bc8/output/x86_64-w64-mingw32/inputs.SHA256SUMS
```

fanquake commented:
> makes it easy to integrate new modules into static builds

Can you demonstrate this is a branch on top of these changes?

hebasto committed:
build, qt: Add linguist_tools list

hebasto committed:
build, qt: Drop lrelease dependency patch

It is no longer required after switching to Qt top-level build.

hebasto committed:
build, qt: Drop translations.pro hack

It is no longer required after switching to Qt top-level build.

hebasto committed:
build, qt: Force bootstrap while building linguist tools

Qt lrelease tool depends on the xml module. This change guarantees that
it is always available after being bootstrapped.

hebasto committed:
build, qt: Fix wrong cross-compiling detection on macOS

hebasto commented:
Updated 5ca6e8f16bc8bf7d4a388aeb8677ba2262d91b8e -> 763793b60e4ec0d1df129279ca3f08fc97d6d90e ([pr20641.05](https://github.com/hebasto/bitcoin/commits/pr20641.05) -> [pr20641.06](https://github.com/hebasto/bitcoin/commits/pr20641.06)):
- dropped some unimportant commits
- improved compatibility with possible module addition
- rebased

@fanquake 

> > makes it easy to integrate new modules into static builds
> 
> Can you demonstrate this in a branch on top of these changes?

See: [hebasto:201213-top-DEMO](https://github.com/hebasto/bitcoin/commits/201213-top-DEMO).

hebasto committed:
build, qt: Do not install *.prl files

hebasto commented:
#### Guix builds:

```
$ find guix-build-$(git rev-parse --short=12 HEAD)/output/ -type f -print0 | env LC_ALL=C sort -z | xargs -r0 sha256sum
5bb23a2c9ff259da9516f763afaec4ebef92214bb21d5f8eb4818019e12d19df  guix-build-1155978d8f3f/output/aarch64-linux-gnu/bitcoin-1155978d8f3f-aarch64-linux-gnu-debug.tar.gz
764d717839128d1a5194b48a953ee162248a782d8614a5aa8f498c27460768f1  guix-build-1155978d8f3f/output/aarch64-linux-gnu/bitcoin-1155978d8f3f-aarch64-linux-gnu.tar.gz
c33129ffcec4e1ced5b21663facc034f08ffcb7e5992ec1e1916e61d2e500a1f  guix-build-1155978d8f3f/output/aarch64-linux-gnu/inputs.SHA256SUMS
0e3bd95c99ba0da448fd9472044b914170e6b9dc4ac7e2b2c44925b9d69bda24  guix-build-1155978d8f3f/output/arm-linux-gnueabihf/bitcoin-1155978d8f3f-arm-linux-gnueabihf-debug.tar.gz
9d47357c5f17d06299fa8e0185738a824be77b6bf1436f478dc0a6ed82c7b322  guix-build-1155978d8f3f/output/arm-linux-gnueabihf/bitcoin-1155978d8f3f-arm-linux-gnueabihf.tar.gz
c33129ffcec4e1ced5b21663facc034f08ffcb7e5992ec1e1916e61d2e500a1f  guix-build-1155978d8f3f/output/arm-linux-gnueabihf/inputs.SHA256SUMS
e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855  guix-build-1155978d8f3f/output/dist-archive/SKIPATTEST.TAG
13ab7d91d5b1f74d0c78eec7a883936b06bfaae588f7b1526fa90217ca8c036d  guix-build-1155978d8f3f/output/dist-archive/bitcoin-1155978d8f3f.tar.gz
b41233eb7319591fe2cb4e672e043dafdb72b07ee222e7fe9325f1ea017a78ee  guix-build-1155978d8f3f/output/powerpc64-linux-gnu/bitcoin-1155978d8f3f-powerpc64-linux-gnu-debug.tar.gz
567684b9b557faf75995974f1c5ebfd416404041ada78681802b975375a7c28f  guix-build-1155978d8f3f/output/powerpc64-linux-gnu/bitcoin-1155978d8f3f-powerpc64-linux-gnu.tar.gz
c33129ffcec4e1ced5b21663facc034f08ffcb7e5992ec1e1916e61d2e500a1f  guix-build-1155978d8f3f/output/powerpc64-linux-gnu/inputs.SHA256SUMS
598dbdea2a7cf0dea53319dfa51a71cafe1704241fb4212627371091a477aacb  guix-build-1155978d8f3f/output/powerpc64le-linux-gnu/bitcoin-1155978d8f3f-powerpc64le-linux-gnu-debug.tar.gz
a40d830d088d411a0b11e43fd0df8c0369477494ee3d8dd621d6f81784bcf195  guix-build-1155978d8f3f/output/powerpc64le-linux-gnu/bitcoin-1155978d8f3f-powerpc64le-linux-gnu.tar.gz
c33129ffcec4e1ced5b21663facc034f08ffcb7e5992ec1e1916e61d2e500a1f  guix-build-1155978d8f3f/output/powerpc64le-linux-gnu/inputs.SHA256SUMS
dd8591dc616fe54448078cad962704ba226994aa232f65bc3db7312d68eb15c8  guix-build-1155978d8f3f/output/riscv64-linux-gnu/bitcoin-1155978d8f3f-riscv64-linux-gnu-debug.tar.gz
7c63e4f88eee0692ed206cfe8724b50f7ee81844a433f30254a1e7afce16e8c2  guix-build-1155978d8f3f/output/riscv64-linux-gnu/bitcoin-1155978d8f3f-riscv64-linux-gnu.tar.gz
c33129ffcec4e1ced5b21663facc034f08ffcb7e5992ec1e1916e61d2e500a1f  guix-build-1155978d8f3f/output/riscv64-linux-gnu/inputs.SHA256SUMS
29c151d971ae7cf1ebaf90a29882f32a777d4221dbc8e1ffb75c87e66301cbbe  guix-build-1155978d8f3f/output/x86_64-apple-darwin18/bitcoin-1155978d8f3f-osx-unsigned.dmg
335a5f02fed8487635924ffcd8cd60d705b090a31c543e4d5a2adb3538dcf333  guix-build-1155978d8f3f/output/x86_64-apple-darwin18/bitcoin-1155978d8f3f-osx-unsigned.tar.gz
932925c550c588f5ebff562ee5c495fbe437cd7e2a620aa965ca4f578c5baf1d  guix-build-1155978d8f3f/output/x86_64-apple-darwin18/bitcoin-1155978d8f3f-osx64.tar.gz
c33129ffcec4e1ced5b21663facc034f08ffcb7e5992ec1e1916e61d2e500a1f  guix-build-1155978d8f3f/output/x86_64-apple-darwin18/inputs.SHA256SUMS
5e30a07ed4976a6cbe1ca24b3c3e03954c65132a2cdc9d7d6bbc0e3a18521600  guix-build-1155978d8f3f/output/x86_64-linux-gnu/bitcoin-1155978d8f3f-x86_64-linux-gnu-debug.tar.gz
d602eaf5d69bc3c29d28a6540b565cf330d0905fcf365752989757fd5b6d5990  guix-build-1155978d8f3f/output/x86_64-linux-gnu/bitcoin-1155978d8f3f-x86_64-linux-gnu.tar.gz
c33129ffcec4e1ced5b21663facc034f08ffcb7e5992ec1e1916e61d2e500a1f  guix-build-1155978d8f3f/output/x86_64-linux-gnu/inputs.SHA256SUMS
c98df448196f382d8b4f0046210535e2ec75cc70161cf248e91f8e75f075de23  guix-build-1155978d8f3f/output/x86_64-w64-mingw32/bitcoin-1155978d8f3f-win-unsigned.tar.gz
10a3b9196678ae10229f99bf74a10119c82e665eec21d1c5b77c81ba758803c2  guix-build-1155978d8f3f/output/x86_64-w64-mingw32/bitcoin-1155978d8f3f-win64-debug.zip
6b2bdd460c6df7ef5dcf149feae5c92056713a6f173bb45737c2c95ccc89efd1  guix-build-1155978d8f3f/output/x86_64-w64-mingw32/bitcoin-1155978d8f3f-win64-setup-unsigned.exe
7ba0334921fffa8d9989d1d110c0bb106049362f562e5d3bc6e2cdfa642c4084  guix-build-1155978d8f3f/output/x86_64-w64-mingw32/bitcoin-1155978d8f3f-win64.zip
c33129ffcec4e1ced5b21663facc034f08ffcb7e5992ec1e1916e61d2e500a1f  guix-build-1155978d8f3f/output/x86_64-w64-mingw32/inputs.SHA256SUMS
```

hebasto commented:
#### Gitian builds:

- Linux
```
Generating report
1d5173dedbcac24a458b97bdb8e2634dc8434e40ab84d5c551d3a4fb680d0344  bitcoin-1155978d8f3f-aarch64-linux-gnu-debug.tar.gz
9b05b71039bc0690bd6d44342f1823354acc3567098c1a1e647e93a27cc37459  bitcoin-1155978d8f3f-aarch64-linux-gnu.tar.gz
865f31945c0f1519b4fb0b3401ad54d606c11ba38326eaef2b61d3c5fd9c7972  bitcoin-1155978d8f3f-arm-linux-gnueabihf-debug.tar.gz
5c070025e7335a21cbb6248aa270c90981657666339b186b7d407c46e9c6be89  bitcoin-1155978d8f3f-arm-linux-gnueabihf.tar.gz
8fec4d8bcfab16c0e400bd8bf9e1a1ac8755b6333c97b2423b3ee3bac272a4d7  bitcoin-1155978d8f3f-powerpc64-linux-gnu-debug.tar.gz
c60243bd0dfe803b107b8a65a390cacfe5b5b63fd5b4cb6f8b0f7cdae26eadd8  bitcoin-1155978d8f3f-powerpc64-linux-gnu.tar.gz
0e617819e5c794c37755dcfbef43acc2e77b321bfda9286c264579b41b161d6d  bitcoin-1155978d8f3f-powerpc64le-linux-gnu-debug.tar.gz
fa4a9f4bd556700be54aaede22a7d75952db1519659493fce213fb97a0cafb5b  bitcoin-1155978d8f3f-powerpc64le-linux-gnu.tar.gz
79791ecf185cb42ace36c4bdd7f6fdc390c519fe6b93850ecd659c790142c213  bitcoin-1155978d8f3f-riscv64-linux-gnu-debug.tar.gz
63026c3389db0c2e6d5c1a244afae8ce715f1acc70477f704b1ff52894a1fe68  bitcoin-1155978d8f3f-riscv64-linux-gnu.tar.gz
9c0e483e570285e0e551b442ea1cf858b66f51363c5ba8abb00a85631341e7f5  bitcoin-1155978d8f3f-x86_64-linux-gnu-debug.tar.gz
15c7aefe3dab9b1aaa2d851cbd7f36652f36adc56f69130ab48f6a387e6b5cdf  bitcoin-1155978d8f3f-x86_64-linux-gnu.tar.gz
13ab7d91d5b1f74d0c78eec7a883936b06bfaae588f7b1526fa90217ca8c036d  src/bitcoin-1155978d8f3f.tar.gz
e9ee4bf4b9b4ccec570859ac7acb3e97b6179a18369c482e1eea2771a7a8052a  bitcoin-core-linux-22-res.yml
Done.
```

- Windows
```
Generating report
cd5c5a76c29e56175232c69a6f10fd52e971648cc13ff8979f31f277072ce714  bitcoin-1155978d8f3f-win-unsigned.tar.gz
6926acdb75d69d53579fde33a488bce2e9c8c57aed70a493ea10d8926c717da3  bitcoin-1155978d8f3f-win64-debug.zip
8f24dc400f14e5f35e3094edf9f5138e70468dff9828b9530f02f35898dc19d0  bitcoin-1155978d8f3f-win64-setup-unsigned.exe
b0244f71bd8877958e8b07e631295d6851d8b3b52e51c6eab2e05cec54722e40  bitcoin-1155978d8f3f-win64.zip
13ab7d91d5b1f74d0c78eec7a883936b06bfaae588f7b1526fa90217ca8c036d  src/bitcoin-1155978d8f3f.tar.gz
a0a679c11578b9132e725a9c24d6253d18b2a2ddb4e2c022ee65406763889e33  bitcoin-core-win-22-res.yml
Done.
```

- macOS
```
Generating report
006c76657ed2bcd17b39f6c1973c8b00dc0c31c3c2e9799ce4ca1c29518e997b  bitcoin-1155978d8f3f-osx-unsigned.dmg
8523361566c2fcf06d07eeaa58fb6b1bac38e23927ab3c54612e59e9ef17f17f  bitcoin-1155978d8f3f-osx-unsigned.tar.gz
d7b0ab0f32709639ce16725d90adb9bb83c280d9aa4628de3b1cd2b226564524  bitcoin-1155978d8f3f-osx64.tar.gz
13ab7d91d5b1f74d0c78eec7a883936b06bfaae588f7b1526fa90217ca8c036d  src/bitcoin-1155978d8f3f.tar.gz
cdc4a70d5be5767572c1081092b9e3a285e018b2ca89842c5fcb72b3b7ada53d  bitcoin-core-osx-22-res.yml
Done.
```

hebasto commented:
@fanquake
 
> > Can you demonstrate this in a branch on top of these changes?
> 
> See: [hebasto:201213-top-DEMO](https://github.com/hebasto/bitcoin/commits/201213-top-DEMO).

CI on Cirrus: https://cirrus-ci.com/build/5900358854639616

The only [error](https://cirrus-ci.com/task/5275098656538624) is about linking against `xcb-shm` (see #21376 and #22174), but it is not related to this PR changes directly.

fanquake commented:
Guix builds:
```bash
5bb23a2c9ff259da9516f763afaec4ebef92214bb21d5f8eb4818019e12d19df  guix-build-1155978d8f3f/output/aarch64-linux-gnu/bitcoin-1155978d8f3f-aarch64-linux-gnu-debug.tar.gz
764d717839128d1a5194b48a953ee162248a782d8614a5aa8f498c27460768f1  guix-build-1155978d8f3f/output/aarch64-linux-gnu/bitcoin-1155978d8f3f-aarch64-linux-gnu.tar.gz
c33129ffcec4e1ced5b21663facc034f08ffcb7e5992ec1e1916e61d2e500a1f  guix-build-1155978d8f3f/output/aarch64-linux-gnu/inputs.SHA256SUMS
0e3bd95c99ba0da448fd9472044b914170e6b9dc4ac7e2b2c44925b9d69bda24  guix-build-1155978d8f3f/output/arm-linux-gnueabihf/bitcoin-1155978d8f3f-arm-linux-gnueabihf-debug.tar.gz
9d47357c5f17d06299fa8e0185738a824be77b6bf1436f478dc0a6ed82c7b322  guix-build-1155978d8f3f/output/arm-linux-gnueabihf/bitcoin-1155978d8f3f-arm-linux-gnueabihf.tar.gz
c33129ffcec4e1ced5b21663facc034f08ffcb7e5992ec1e1916e61d2e500a1f  guix-build-1155978d8f3f/output/arm-linux-gnueabihf/inputs.SHA256SUMS
e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855  guix-build-1155978d8f3f/output/dist-archive/SKIPATTEST.TAG
13ab7d91d5b1f74d0c78eec7a883936b06bfaae588f7b1526fa90217ca8c036d  guix-build-1155978d8f3f/output/dist-archive/bitcoin-1155978d8f3f.tar.gz
b41233eb7319591fe2cb4e672e043dafdb72b07ee222e7fe9325f1ea017a78ee  guix-build-1155978d8f3f/output/powerpc64-linux-gnu/bitcoin-1155978d8f3f-powerpc64-linux-gnu-debug.tar.gz
567684b9b557faf75995974f1c5ebfd416404041ada78681802b975375a7c28f  guix-build-1155978d8f3f/output/powerpc64-linux-gnu/bitcoin-1155978d8f3f-powerpc64-linux-gnu.tar.gz
c33129ffcec4e1ced5b21663facc034f08ffcb7e5992ec1e1916e61d2e500a1f  guix-build-1155978d8f3f/output/powerpc64-linux-gnu/inputs.SHA256SUMS
598dbdea2a7cf0dea53319dfa51a71cafe1704241fb4212627371091a477aacb  guix-build-1155978d8f3f/output/powerpc64le-linux-gnu/bitcoin-1155978d8f3f-powerpc64le-linux-gnu-debug.tar.gz
a40d830d088d411a0b11e43fd0df8c0369477494ee3d8dd621d6f81784bcf195  guix-build-1155978d8f3f/output/powerpc64le-linux-gnu/bitcoin-1155978d8f3f-powerpc64le-linux-gnu.tar.gz
c33129ffcec4e1ced5b21663facc034f08ffcb7e5992ec1e1916e61d2e500a1f  guix-build-1155978d8f3f/output/powerpc64le-linux-gnu/inputs.SHA256SUMS
dd8591dc616fe54448078cad962704ba226994aa232f65bc3db7312d68eb15c8  guix-build-1155978d8f3f/output/riscv64-linux-gnu/bitcoin-1155978d8f3f-riscv64-linux-gnu-debug.tar.gz
7c63e4f88eee0692ed206cfe8724b50f7ee81844a433f30254a1e7afce16e8c2  guix-build-1155978d8f3f/output/riscv64-linux-gnu/bitcoin-1155978d8f3f-riscv64-linux-gnu.tar.gz
c33129ffcec4e1ced5b21663facc034f08ffcb7e5992ec1e1916e61d2e500a1f  guix-build-1155978d8f3f/output/riscv64-linux-gnu/inputs.SHA256SUMS
29c151d971ae7cf1ebaf90a29882f32a777d4221dbc8e1ffb75c87e66301cbbe  guix-build-1155978d8f3f/output/x86_64-apple-darwin18/bitcoin-1155978d8f3f-osx-unsigned.dmg
335a5f02fed8487635924ffcd8cd60d705b090a31c543e4d5a2adb3538dcf333  guix-build-1155978d8f3f/output/x86_64-apple-darwin18/bitcoin-1155978d8f3f-osx-unsigned.tar.gz
932925c550c588f5ebff562ee5c495fbe437cd7e2a620aa965ca4f578c5baf1d  guix-build-1155978d8f3f/output/x86_64-apple-darwin18/bitcoin-1155978d8f3f-osx64.tar.gz
c33129ffcec4e1ced5b21663facc034f08ffcb7e5992ec1e1916e61d2e500a1f  guix-build-1155978d8f3f/output/x86_64-apple-darwin18/inputs.SHA256SUMS
5e30a07ed4976a6cbe1ca24b3c3e03954c65132a2cdc9d7d6bbc0e3a18521600  guix-build-1155978d8f3f/output/x86_64-linux-gnu/bitcoin-1155978d8f3f-x86_64-linux-gnu-debug.tar.gz
d602eaf5d69bc3c29d28a6540b565cf330d0905fcf365752989757fd5b6d5990  guix-build-1155978d8f3f/output/x86_64-linux-gnu/bitcoin-1155978d8f3f-x86_64-linux-gnu.tar.gz
c33129ffcec4e1ced5b21663facc034f08ffcb7e5992ec1e1916e61d2e500a1f  guix-build-1155978d8f3f/output/x86_64-linux-gnu/inputs.SHA256SUMS
c98df448196f382d8b4f0046210535e2ec75cc70161cf248e91f8e75f075de23  guix-build-1155978d8f3f/output/x86_64-w64-mingw32/bitcoin-1155978d8f3f-win-unsigned.tar.gz
10a3b9196678ae10229f99bf74a10119c82e665eec21d1c5b77c81ba758803c2  guix-build-1155978d8f3f/output/x86_64-w64-mingw32/bitcoin-1155978d8f3f-win64-debug.zip
6b2bdd460c6df7ef5dcf149feae5c92056713a6f173bb45737c2c95ccc89efd1  guix-build-1155978d8f3f/output/x86_64-w64-mingw32/bitcoin-1155978d8f3f-win64-setup-unsigned.exe
7ba0334921fffa8d9989d1d110c0bb106049362f562e5d3bc6e2cdfa642c4084  guix-build-1155978d8f3f/output/x86_64-w64-mingw32/bitcoin-1155978d8f3f-win64.zip
c33129ffcec4e1ced5b21663facc034f08ffcb7e5992ec1e1916e61d2e500a1f  guix-build-1155978d8f3f/output/x86_64-w64-mingw32/inputs.SHA256SUMS
```

fanquake reviewed (approved):
ACK 1155978d8f3fcc1cebf357302b933b834f9c9465
